loki/transformations/single_column/vertical.py

```diff
@@ -0,0 +1,218 @@
+        return ignore_names
+
+    def identify_multilevel_arrays(self, local_arrays):
+        """
+        Identify local arrays/temporaries that have an access in the vertical dimension
+        that is different to '<vertical.index>', e.g., '<vertical.index> +/- 1'
+        """
+        multilevel_local_arrays = []
+        for arr in local_arrays:
+            for dim in arr.dimensions:
+                if self.vertical.index in FindVariables().visit(dim):
+                    # dim is not equal to vertical.index e.g., vertical.index +/- 1
+                    if dim != self.vertical.index:
+                        multilevel_local_arrays.append(arr)
+        return multilevel_local_arrays
+
+    def correct_init_of_multilevel_arrays(self, routine, multilevel_local_arrays):
+        """
+        Possibly handle initaliztion of those multilevel local arrays via
+        splitting relevant loops or rather creating a new node with the relevant
+        nodes moved to the newly created loop.
+
+        .. note::
+            This relies on pragmas being inserted in the input source code!
+        """
+        loop_map = {}
+        # find/identify loops with pragma 'loop-fusion group(<group-name>-init)'
+        with pragmas_attached(routine, ir.Loop):
+            loop_map = {}
+            for loop in FindNodes(ir.Loop).visit(routine.body):
+                if is_loki_pragma(loop.pragma, starts_with='loop-fusion'):
+                    parameters = get_pragma_parameters(loop.pragma, starts_with='loop-fusion')
+                    group = parameters.get('group', 'default')
+                    if 'init' in group:
+                        nodes_to_be_moved = ()
+                        nodes = FindNodes(ir.Assignment).visit(loop.body)
+                        node_map = {}
+                        node_map_init = {}
+                        # find nodes that have multilevel arrays
+                        for node in nodes:
+                            node_vars = FindVariables().visit(node)
+                            if any(node_var.name.lower() in multilevel_local_arrays for node_var in node_vars):
+                                nodes_to_be_moved += (node,)
+                                node_map[node] = None
+                            else:
+                                node_map_init[node] = None
+                        # split the loop/create a new node to move those nodes with
+                        # multilevel arrays to the new node
+                        if nodes_to_be_moved:
+                            pragmas = loop.pragma
+                            new_pragmas = [pragma.clone(content=pragma.content.replace('-init', '')) if '-init'
+                                    in pragma.content else pragma for pragma in pragmas]
+                            loop_map[loop] = (ir.Comment('! Loki generated loop for init ...'),
+                                    Transformer(node_map_init).visit(loop.clone(\
```

Comment:
Two inline `Transformer` invocations makes this very hard to read. Can we apply them in sequence to build the tuple, before putting them into the `loop_map`?

Reply:
Not sure what you mean.
I need to apply the Transformer twice with two different maps since the key would be overwritten otherwise ...
If you propose to introduce temporary variables: sure this is possible!

Reply:
Oh, I just meant the cosmetics of it - the use of two transformers is fine, it just gets really hard to read if everything is applied in a single line. If you build the tuple incrementally, each `Transformer.visit(..)` invocation could be a single line - but this is really just a cosmetic suggestion and not strictly necessary.